---
title: "Differential Expression Analysis: visualization"
date: "Created: 2021-08-17 Updated: `r Sys.Date()`"
output: 
  html_notebook:
    codes: hide
---

### Introdution
```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = F, message = F)
library(dplyr)
library(tibble)
library(convert)
library(ggplot2)
library(ggrepel)
library(data.table)

# rm(list = ls())
options(stringsAsFactors = F)
options(future.globals.maxSize = 1000 * 1024^2)

# rm(list = ls())
options(stringsAsFactors = F)
options(future.globals.maxSize = 1000 * 1024^2)

grp <- c("Normal", "Tumor")
grp.col <- c("#568875", "#73FAFC")
```


### Importing data 
```{r}
HNSC_ExprSet <- readRDS("../../Study/SeqData/TCGA_HNSC/Clean/TCGA-HNSC-post_mRNA_ExprSet.RDS")
HNSC_DEG <- fread("../../Result/Differential/HNSC_limma_gene.csv")
table(HNSC_DEG$Enrichment)
```


### Volcano Function
```{r, fig.width=8, fig.height=6}
volcanofun <- function(datset=HNSC_DEG,
                       genelist=HNSC_DEG$GeneID[1:10],
                       group_name=grp,
                       group_col=grp.col,
                       pval=0.05, 
                       fc=0.6){
  
  # datset=HNSC_DEG
  # genelist=HNSC_DEG$GeneID[1:10]
  # group_name=grp
  # group_col=grp.col
  # pval=0.05
  # fc=0.6
  
  dat <- datset %>% 
    mutate(color=factor(Enrichment, levels = c(group_name, "Nonsignif")))  
  # print(table(dat$color))
  dat_status <- table(dat$color)
  dat_status_number <- as.numeric(dat_status)
  dat_status_name <- names(dat_status)
  legend_label <- c(paste0(dat_status_name[1], " (", dat_status_number[1], ")"),
                    paste0(dat_status_name[2], " (", dat_status_number[2], ")"),
                    paste0("Nonsignif", " (", dat_status_number[3], ")"))
  
  dat.signif <- subset(dat, Enrichment != "Nonsignif") %>%
    filter(GeneID%in%genelist)
  print(table(dat.signif$color))
  
  group_col_new <- c(group_col, "grey")
  group_name_new <- levels(dat$color)
  
  xlabel <- paste0("log2(FC) [", paste(group_name, collapse=":"), "]")
  
  # Make a basic ggplot2 object with x-y values
  pl <- ggplot(dat, aes(x = -logFC, y = -log10(adj.P.Val), color = color))+ 
          geom_point(size = 0.8, alpha = 0.8)+
          scale_color_manual(name = NULL,
                             values = group_col_new,
                             labels = c(legend_label, "Nonsignif"))+
          xlab(xlabel) + 
          ylab(expression(-log[10]("adjusted p-value")))+ 
          geom_hline(yintercept=-log10(pval), alpha=.8, linetype=2, size=.7)+
          geom_vline(xintercept=fc, alpha=.8, linetype=2, size=.7)+
          geom_vline(xintercept=-fc, alpha=.8, linetype=2, size=.7)+ 
          geom_text_repel(data = dat.signif,
                          aes(label = GeneID),
                          size = 4,
                          max.overlaps = getOption("ggrepel.max.overlaps", default = 80),
                          segment.linetype = 1,
                          segment.curvature = -1e-20,
                          box.padding = unit(0.35, "lines"),
                          point.padding = unit(0.3, "lines"),
                          arrow = arrow(length = unit(0.005, "npc")),
                          color = "black",     # text color
                          bg.color = "white", # shadow color
                          bg.r = 0.15)+
          annotate("text", x=min(dat$logFC), y=-log10(pval), label=pval, size=6, color="red")+
          annotate("text", x=fc, y=0, label=fc, size=6, color="red")+
          annotate("text", x=-fc, y=0, label=-fc, size=6, color="red")+
          scale_y_continuous(trans = "log1p")+
          guides(color=guide_legend(override.aes = list(size = 3)))+
          theme_bw()+ 
          theme(axis.title = element_text(color = "black", size = 12),
                axis.text = element_text(color = "black", size = 10),
                text = element_text(size = 8, color = "black", family="serif"),
                panel.grid = element_blank(),
                #legend.position = "right",
                legend.position = c(.15, .1),
                legend.key.height = unit(0.6,"cm"),
                legend.text = element_text(face = "bold", color = "black", size = 8),
                strip.text = element_text(face = "bold", size = 14))
  return(pl)
}

Gene_boxplot <- function(
        datset=HNSC_ExprSet,
        genelist=HNSC_DEG$GeneID[1:10],
        group_name=grp,
        group_col=grp.col){

  # datset=HNSC_ExprSet
  # genelist=HNSC_DEG$GeneID[1:10]
  # group_name=grp
  # group_col=grp.col
  
  pheno <- pData(datset) %>%
    filter(Group %in% group_name)
  pheno$Group <- factor(as.character(pheno$Group), levels = group_name)
  
  edata <- data.frame(exprs(datset)) %>%
    rownames_to_column("GeneID") %>%
    filter(GeneID%in%genelist) %>%
    column_to_rownames("GeneID")
  edata <- edata[, colnames(edata)%in%rownames(pheno)]
  
  mdat <- pheno %>% dplyr::select(Group) %>%
    rownames_to_column("SampleID") %>%
    inner_join(t(edata) %>% data.frame() %>% rownames_to_column("SampleID"), by = "SampleID") %>%
    column_to_rownames("SampleID")
  plotdata <- mdat %>% tidyr::gather(key="geneID", value="value", -c("Group")) 
  
  # arrange by median of group per genes 
  # plotdata_median_delta <- plotdata %>% group_by(Type, geneID) %>%
  #   summarise(median_value=median(value)) %>%
  #   ungroup() %>%
  #   arrange(geneID, Type, median_value) %>%
  #   group_by(geneID) %>%
  #   summarise(delta=dplyr::first(median_value) - dplyr::last(median_value)) %>%
  #   arrange(delta)
  
  plotdata$geneID <- factor(plotdata$geneID, levels = genelist)  
  
  pl <- ggplot(plotdata, aes(x = Group, y = value, fill= Group))+
    stat_boxplot(geom = "errorbar", width = 0.15,
                 position = position_dodge(0.4)) +    
    geom_boxplot(width = 0.4, 
                 outlier.colour = "black", 
                 outlier.shape=21, outlier.size = 1)+
    scale_fill_manual(values = group_col)+
    scale_y_continuous(labels = scales::scientific)+
    facet_wrap(facets = "geneID", scales = "free_y", nrow = 2)+    
    labs(x="", y="Gene TPM")+
    guides(fill=F)+
    theme_classic()+
    theme(axis.title = element_text(color = "black", size = 12),
          axis.text.x = element_text(color = "black", size = 10, hjust = .5, vjust = .5),
          text = element_text(size = 8, color = "black", family="serif"),
          panel.grid = element_blank(),
          strip.text = element_text(face = "bold", size = 12))    
    
  return(pl)
}


HNSC_volcano <- volcanofun(datset=HNSC_DEG)
HNSC_volcano
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/HNSC_volcano.pdf", HNSC_volcano, width = 8, height = 6)

HNSC_boxplot <- Gene_boxplot(datset=HNSC_ExprSet)
HNSC_boxplot
if(!dir.exists("../../Result/Figure")){
  dir.create("../../Result/Figure")
}
ggsave("../../Result/Figure/HNSC_boxplot.pdf", HNSC_boxplot, width = 8, height = 6)
```


### Heatmap Of DEGs 
```{r, fig.width=10, fig.height=10}
heatFun <- function(datset=HNSC_DEG,
                    thresholdFC=4,
                    ExprSet=HNSC_ExprSet,
                    group_name=grp){
  
  # datset=HNSC_DEG
  # thresholdFC=4
  # ExprSet=HNSC_ExprSet
  # group_name=grp
  
  
  diff_gene <- datset %>% filter(Enrichment != "Nonsignif") %>%
    filter(abs(logFC) > thresholdFC)
  
  pheno <- pData(ExprSet) %>% data.frame() %>%
    rownames_to_column("Barcode") %>%
    filter(Group%in%group_name) %>%
    mutate(Group=factor(Group, levels = group_name)) %>%
    arrange(Group) %>%
    column_to_rownames("Barcode")
  
  edata <- exprs(ExprSet) %>% data.frame() %>%
    rownames_to_column("geneid") %>%
    filter(geneid%in%diff_gene$GeneID) %>%
    dplyr::select(c("geneid", rownames(pheno))) %>%
    column_to_rownames("geneid")
  
  # scale data: z-score
  scale_rows <- function (x) {
      m = apply(x, 1, mean, na.rm = T)
      s = apply(x, 1, sd, na.rm = T)
      return((x - m)/s)
  }  
  edata_scaled <- t(scale_rows(edata))
  require(circlize)
  col_fun <- colorRamp2(c(round(range(edata_scaled)[1]), 0, 
                          round(range(edata_scaled)[2])),
                        c("blue", "black", "red")) 
  # row split 
  dat_status <- table(pheno$Group)
  dat_status_number <- as.numeric(dat_status)
  dat_status_name <- names(dat_status)
  row_split <- c()
  for (i in 1:length(dat_status_number)) {
    row_split <- c(row_split, rep(i, dat_status_number[i]))
  }
  require(ComplexHeatmap)
  Heatmap(
          edata_scaled, 
          #col = col_fun,
          cluster_rows = FALSE,
          row_order = rownames(pheno),
          show_column_names = FALSE,
          show_row_names = FALSE,
          row_names_gp = gpar(fontsize = 12),
          row_names_side = "right",
          row_dend_side = "left",
          column_title = NULL, 
          heatmap_legend_param = list(
            title = "Gene Expression\nZscore",
            title_position = "topcenter",
            border = "black",
            legend_height = unit(10, "cm"),
            direction = "horizontal"),
         row_split = row_split,
        left_annotation = rowAnnotation(foo = anno_block(gp = gpar(fill = 2:4),
            labels = group_name, 
            labels_gp = gpar(col = "black", fontsize = 12))),         
         column_km = 3
    )
}

heatFun(datset=HNSC_DEG,
        thresholdFC=4,
        ExprSet=HNSC_ExprSet,
        group_name=grp)
```


### systemic information
```{r}
sessionInfo()
```


### Reference

1. [the definition of limma logFC](https://kasperdanielhansen.github.io/genbioconductor/html/limma.html)
